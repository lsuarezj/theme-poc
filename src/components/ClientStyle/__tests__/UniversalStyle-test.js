import React, { Component } from "react";
import { render } from "react-dom";
import { renderToString } from "react-dom/server";
import { JSDOM } from "jsdom";

import Style from "../UniversalStyle";

afterEach(() => {
  global.window = undefined;
  global.document = undefined;
});

describe("UniversalComponent Component", () => {
  it("should render a style element on the server", () => {
    const css = ".class1 { color: red } .class2 { color: blue }";

    const view = renderToString(<Style css={css} />);
    expect(view).toMatchSnapshot();
  });

  it("should render a style element on the client", () => {
    const { window } = new JSDOM("<html><body></body></html>");

    global.window = window;
    global.document = window.document;

    const div = document.createElement("div");
    const css = ".class1 { color: red } .class2 { color: blue }";

    render(<Style css={css} />, div);

    expect(div.innerHTML).toMatchSnapshot();
    expect(document.head.childNodes.length).toBe(0);
  });

  it("should move the style to the document.head if the component unmounts", () => {
    const { window } = new JSDOM("<html><body></body></html>");

    global.window = window;
    global.document = window.document;

    const div = document.createElement("div");
    const css = ".class1 { color: red } .class2 { color: blue }";

    class Wrapper extends Component {
      constructor(props, context) {
        super(props, context);

        this.state = {
          isVisible: true,
        };

        window.hide = () => {
          this.setState({ isVisible: false });
        };
      }

      render() {
        if (this.state.isVisible) {
          return <Style css={css} />;
        }

        return null;
      }
    }

    render(<Wrapper />, div);

    expect(div.innerHTML).toMatchSnapshot();
    expect(document.head.childNodes.length).toBe(0);

    window.hide();

    expect(div.innerHTML).toBe("");
    expect(document.head.childNodes.length).toBe(1);
    expect(
      document.getElementById("__react_css_component_id-0").textContent
    ).toBe(css);
  });

  it("should move each different style to the document.head if the component unmounts", () => {
    const { window } = new JSDOM("<html><body></body></html>");

    global.window = window;
    global.document = window.document;

    const div = document.createElement("div");
    const css1 = ".class1 { color: red } .class2 { color: blue }";
    const css2 = ".class3 { color: green }";

    class Wrapper extends Component {
      constructor(props, context) {
        super(props, context);

        this.state = {
          isVisible: true,
        };

        window.hide = () => {
          this.setState({ isVisible: false });
        };
      }

      render() {
        if (this.state.isVisible) {
          return [
            <Style key="style-1" css={css1} />,
            <Style key="style-2" css={css2} />,
          ];
        }

        return null;
      }
    }

    render(<Wrapper />, div);

    expect(div.innerHTML).toMatchSnapshot();
    expect(document.head.childNodes.length).toBe(0);

    window.hide();

    expect(div.innerHTML).toBe("");
    expect(document.head.childNodes.length).toBe(2);
    expect(
      document.getElementById("__react_css_component_id-0").textContent
    ).toBe(css1);
    expect(
      document.getElementById("__react_css_component_id-1").textContent
    ).toBe(css2);
  });

  it("should only move the style once to the document.head", () => {
    const { window } = new JSDOM("<html><body></body></html>");

    global.window = window;
    global.document = window.document;

    const div = document.createElement("div");
    const css = ".class1 { color: red } .class2 { color: blue }";

    class Wrapper extends Component {
      constructor(props, context) {
        super(props, context);

        this.state = {
          isVisible: true,
        };

        window.hide = () => {
          this.setState({ isVisible: false });
        };
      }

      render() {
        if (this.state.isVisible) {
          return [
            <Style key="style-1" css={css} />,
            <Style key="style-2" css={css} />,
          ];
        }

        return null;
      }
    }

    render(<Wrapper />, div);

    expect(div.innerHTML).toMatchSnapshot();
    expect(document.head.childNodes.length).toBe(0);

    window.hide();

    expect(div.innerHTML).toBe("");
    expect(document.head.childNodes.length).toBe(1);
    expect(
      document.getElementById("__react_css_component_id-0").textContent
    ).toBe(css);
  });
});
