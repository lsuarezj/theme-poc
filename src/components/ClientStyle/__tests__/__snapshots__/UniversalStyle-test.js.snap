// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniversalComponent Component should move each different style to the document.head if the component unmounts 1`] = `"<style>.class1 { color: red } .class2 { color: blue }</style><style>.class3 { color: green }</style>"`;

exports[`UniversalComponent Component should move the style to the document.head if the component unmounts 1`] = `"<style>.class1 { color: red } .class2 { color: blue }</style>"`;

exports[`UniversalComponent Component should only move the style once to the document.head 1`] = `"<style>.class1 { color: red } .class2 { color: blue }</style><style>.class1 { color: red } .class2 { color: blue }</style>"`;

exports[`UniversalComponent Component should render a style element on the client 1`] = `"<style>.class1 { color: red } .class2 { color: blue }</style>"`;

exports[`UniversalComponent Component should render a style element on the server 1`] = `"<style>.class1 { color: red } .class2 { color: blue }</style>"`;
